### PR TITLE
Con 2206 tracking subscriptions

### PR DESCRIPTION
**Pull request by Carsten89**

### CODE REVIEW

File: streams/pages.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+const ignoredPaths = [
+  SEARCH_PATH,
+  CATEGORY_PATH,
+  ITEM_PATH,
+];
+
+/**
+ * Emits when one of the tracked paths is entered.
+ */
+const enterThemPages$ = routeDidChange$.filter(({ pathname }) => (
```
Comment (richardgorman) [inline streams/pages.js]: Consider a less hipster naming convention.

File: streams/pages.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+import { routeDidChange$ } from '@shopgate/pwa-common/streams/history';
+import { SEARCH_PATH } from '@shopgate/pwa-common-commerce/search/constants';
+import { CATEGORY_PATH } from '@shopgate/pwa-common-commerce/category/constants';
+import { ITEM_PATH } from '@shopgate/pwa-common-commerce/product/constants';
+
+/**
+ * A list of all route paths that should be tracked,
+ * except those that are handled individualy.
```
Comment (richardgorman) [inline streams/pages.js]: individually

File: streams/product.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  ITEM_PATH,
+} from '@shopgate/pwa-common-commerce/product/constants';
+import {
+  getCurrentProduct,
+  getCurrentBaseProductId,
+  getProductDescription,
+  getProductProperties,
+  getProductShipping,
+} from '@shopgate/pwa-common-commerce/product/selectors/product';
+import {
```
Comment (richardgorman) [inline streams/product.js]: This can be a one-liner.

File: streams/product.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+/**
+ * Emits when product result has been received.
+ */
+export const productReceived$ = main$
+  .filter(({ action }) => action.type === RECEIVE_PRODUCT);
+
+/**
+ * Emits when the current product id changed.
+ */
+const productIdChanged$ = main$
```
Comment (richardgorman) [inline streams/product.js]: Consider renaming to `currentProductIdChanged`?

File: streams/product.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+      return (
+        action.type === SET_PRODUCT_ID &&
+        !!action.productId &&
+        action.productId !== prevId
+      );
+    }
+  );
+
+/**
+ * Emits when specific product data has been received.
```
Comment (richardgorman) [inline streams/product.js]: Please add comments for the two streams below and change this one to be a little more specific.

File: streams/product.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  .filter(({ action }) => action.type === RECEIVE_PRODUCT_DESCRIPTION);
+
+const propertiesReceived$ = main$
+  .filter(({ action }) => action.type === RECEIVE_PRODUCT_PROPERTIES);
+
+const shippingReceived$ = main$
+  .filter(({ action }) => action.type === RECEIVE_PRODUCT_SHIPPING);
+
+/**
+ * Emits when all necessary category data has been received.
```
Comment (richardgorman) [inline streams/product.js]: Maybe a big fat comment outlining what has to happen to make this work would be appreciated.

File: streams/product.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+          // Like the code is now, it won't work for products without properties etc
+          return (
+            getCurrentProduct(state) &&
+            getProductDescription(state) &&
+            getProductProperties(state) &&
+            getProductShipping(state)
+          );
+        })
+    )
+  );
```
Comment (richardgorman) [inline streams/product.js]: Please add a line break after this line.

File: package.json [hunk 8]
```diff
@@ -32,23 +32,35 @@
   "dependencies": {},
   "devDependencies": {
     "@shopgate/eslint-config": "^1.0.2",
+    "@shopgate/pwa-common": "^0.1.0",
+    "@shopgate/pwa-common-commerce": "^0.1.0",
```
Comment (richardgorman) [inline package.json]: Would this repo also not be required as a peer dependency?

Comment (richardgorman) [reply package.json]: Also pwa-core, right?

File: selectors/cart.js [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+import { formatProductData } from './product';
+import { convertPriceToString } from '../helpers';
+
+/**
+ * Reformat product data for addToCart from the store to the format our core expects.
+ * @param {Object} product Product from the store
+ * @param {Object} quantity Quantity of the product
+ * @return {Object}
+ */
+const formatAddToCartProductData = ({ product, quantity }) => ({
```
Comment (richardgorman) [inline selectors/cart.js]: Please move these helper functions to a separate file.

File: subscriptions/pages.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+      getState()
+    )
+  );
+};
+
+/**
+ * Pages tracking subscriptions.
+ * @param {Function} subscribe The subscribe function.
+ */
+export default function pages(subscribe) {
```
Comment (richardgorman) [inline subscriptions/pages.js]: Please don't forget about these console warning.

File: subscriptions/product.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+} from '../selectors/product';
+import getPage from '../selectors/page';
+
+/**
+ * Product tracking subscriptions.
+ * @param {Function} subscribe The subscribe function.
+ */
+export default function product(subscribe) {
+  /**
+   * Gets triggered on product variant change/selection
```
Comment (richardgorman) [inline subscriptions/product.js]: Missing period.

File: subscriptions/product.js [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+  subscribe(variantDidChange$, ({ getState }) => {
+    const state = getState();
+    const variant = getSelectedVariantFormatted(state);
+    const baseProduct = getCurrentBaseProductFormatted(state);
+
+    core.track.variantSelected({ variant, baseProduct });
+  });
+
+  /**
+   * Gets triggered on product pageview
```
Comment (richardgorman) [inline subscriptions/product.js]: Missing period.

**Comment by richardgorman** [comment]
Issues:
- Root Category is not tracked
- Login Success is not tracked
- Pageviews via deeplink behaves inconsistently (category doesn't track)

**Comment by richardgorman** [comment]
Approval was a mistake, please implement changes.

File: streams/pages.js (hunk 1 above)
Comment (Carsten89) [reply streams/pages.js]: done

File: streams/pages.js (hunk 2 above)
Comment (Carsten89) [reply streams/pages.js]: done

File: streams/product.js (hunk 3 above)
Comment (Carsten89) [reply streams/product.js]: done

File: streams/product.js (hunk 4 above)
Comment (Carsten89) [reply streams/product.js]: done

File: streams/product.js (hunk 5 above)
Comment (Carsten89) [reply streams/product.js]: done

File: streams/product.js (hunk 7 above)
Comment (Carsten89) [reply streams/product.js]: done

File: package.json (hunk 8 above)
Comment (Carsten89) [reply package.json]: i'm not sure. flo did this

File: selectors/cart.js (hunk 9 above)
Comment (Carsten89) [reply selectors/cart.js]: done

File: subscriptions/product.js (hunk 11 above)
Comment (Carsten89) [reply subscriptions/product.js]: done

File: subscriptions/product.js (hunk 12 above)
Comment (Carsten89) [reply subscriptions/product.js]: done

**Comment by Carsten89** [comment]
Root Category is not tracked
>> fixed

Login Success is not tracked. 
>> i tested it as part of the checkout and the normal one. In worked fine in both cases.

Pageviews via deeplink behaves inconsistently (category doesn't track)
>> fixed

**Pull request merged**